pytensor/tensor/rewriting/subtensor.py

```diff
@@ -1295,12 +1295,28 @@ def local_inplace_setsubtensor(fgraph, node):
 
 @node_rewriter([AdvancedIncSubtensor1], inplace=True)
 def local_inplace_AdvancedIncSubtensor1(fgraph, node):
-    if isinstance(node.op, AdvancedIncSubtensor1) and not node.op.inplace:
-        new_op = node.op.clone_inplace()
-        new_node = new_op(*node.inputs)
-        copy_stack_trace(node.outputs, new_node)
-        return [new_node]
-    return False
+    if node.op.inplace:
+        return
+
+    x, y, idx = node.inputs
+    if fgraph.has_destroyers([x]):
+        # In this case we can't operate inplace, but if x is just an alloc of zeros
+        # We're better off duplicating it and then acting on it inplace.
+        if (
+            x.owner is not None
+            and isinstance(x.owner.op, Alloc)
+            and all(x.owner.inputs[0].type.broadcastable)
+            and isinstance(x.owner.inputs[0], Constant)
+            and x.owner.inputs[0].unique_value == 0
```

Comment:
Add a `check_is_zero` helper? This exact code block is used above as well.

Reply:
done